src/receiver/google_cloud/worker.rs

```diff
@@ -0,0 +1,234 @@
+    process::Stdio,
+    time::{Duration, Instant},
+};
+use thiserror::Error;
+use tokio::{
+    fs::read_dir,
+    io::AsyncWriteExt,
+    process::Command,
+    sync::{mpsc, watch},
+    time::sleep,
+};
+use tracing::{debug, instrument, trace};
+
+use crate::block::{is_valid_block_file, parse_file, precomputed::PrecomputedBlock};
+
+use super::{bucket_file_from_length, MinaNetwork};
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
+pub struct GoogleCloudBlockWorkerData {
+    pub max_length: u64,
+    pub overlap_num: u64,
+    pub temp_blocks_dir: PathBuf,
+    pub update_freq: Duration,
+    pub network: MinaNetwork,
+    pub bucket: String,
+}
+
+#[derive(Debug, Error, Clone)]
+pub enum GoogleCloudBlockWorkerError {
+    TempBlocksDirIsNotADirectory(PathBuf),
+    IOError(String),
+    BlockParseError(PathBuf, String),
+}
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq, Serialize, Deserialize)]
+pub enum GoogleCloudBlockWorkerCommand {
+    Shutdown,
+    GetWorkerData,
+    SetWorkerData(GoogleCloudBlockWorkerData),
+}
+
+pub struct GoogleCloudBlockWorker {
+    worker_data: GoogleCloudBlockWorkerData,
+    blocks_sender: AsyncHeapProducer<PrecomputedBlock>,
+    error_sender: watch::Sender<Option<GoogleCloudBlockWorkerError>>,
+    command_receiver: mpsc::Receiver<GoogleCloudBlockWorkerCommand>,
+    worker_data_sender: watch::Sender<GoogleCloudBlockWorkerData>,
+}
+
+impl GoogleCloudBlockWorker {
+    pub fn new(
+        worker_data: GoogleCloudBlockWorkerData,
+        blocks_sender: AsyncHeapProducer<PrecomputedBlock>,
+        error_sender: watch::Sender<Option<GoogleCloudBlockWorkerError>>,
+        command_receiver: mpsc::Receiver<GoogleCloudBlockWorkerCommand>,
+        worker_data_sender: watch::Sender<GoogleCloudBlockWorkerData>,
+    ) -> Result<Self, GoogleCloudBlockWorkerError> {
+        if !worker_data.temp_blocks_dir.is_dir() {
+            return Err(GoogleCloudBlockWorkerError::TempBlocksDirIsNotADirectory(
+                worker_data.temp_blocks_dir,
+            ));
+        }
+        Ok(Self {
+            worker_data,
+            blocks_sender,
+            error_sender,
+            command_receiver,
+            worker_data_sender,
+        })
+    }
+
+    #[instrument(skip(self))]
+    pub async fn start_loop(&mut self) -> () {
+        loop {
+            debug!("starting new GoogleCloudBlockWorker work unit");
+            let work_unit_started = Instant::now();
+
+            if let Ok(command) = self.command_receiver.try_recv() {
+                match command {
+                    GoogleCloudBlockWorkerCommand::Shutdown => {
+                        trace!("shutting down GoogleCloudBlockWorker");
+                        if tokio::fs::metadata(&self.worker_data.temp_blocks_dir)
+                            .await
+                            .is_ok()
+                        {
+                            tokio::fs::remove_dir_all(&self.worker_data.temp_blocks_dir)
```

Comment:
huh weird this doesn't work. Had to add it again in `GoogleCloudBlockReceiver::drop`

Reply:
What happens if we eliminate it here?